File: perl/calculator.pl
#!/usr/bin/perl
use strict;
use warnings;

package Calculator::Error;
use base 'Error::Simple';

package Calculator;

sub new {
    my ($class) = @_;
    my $self = {
        variables => {}
    };
    bless $self, $class;
    return $self;
}

sub tokenize {
    my ($self, $input) = @_;
    my @tokens;
    my $pos = 0;
    my $length = length($input);
    
    while ($pos < $length) {
        my $char = substr($input, $pos, 1);
        
        if ($char =~ /\s/) {
            $pos++;
            next;
        }
        
        if ($char =~ /[+\-*\/()=]/) {
            push @tokens, $char;
            $pos++;
            next;
        }
        
        if ($char =~ /[a-zA-Z_]/) {
            my $identifier = '';
            while ($pos < $length && substr($input, $pos, 1) =~ /[a-zA-Z0-9_]/) {
                $identifier .= substr($input, $pos, 1);
                $pos++;
            }
            push @tokens, $identifier;
            next;
        }
        
        if ($char =~ /[0-9]/) {
            my $number = '';
            while ($pos < $length && substr($input, $pos, 1) =~ /[0-9.]/) {
                $number .= substr($input, $pos, 1);
                $pos++;
            }
            if ($number =~ /^\d+\.?\d*$/) {
                push @tokens, $number;
            } else {
                Calculator::Error->throw("Geçersiz sayı formatı: $number");
            }
            next;
        }
        
        Calculator::Error->throw("Geçersiz karakter: $char");
    }
    
    return \@tokens;
}

sub evaluate {
    my ($self, $tokens) = @_;
    return 0 unless @$tokens;
    
    my $token = $tokens->[0];
    
    if ($token eq 'print') {
        if (@$tokens < 2) {
            Calculator::Error->throw("Print ifadesi için değer eksik");
        }
        my $var_name = $tokens->[1];
        if (!exists $self->{variables}{$var_name}) {
            Calculator::Error->throw("Tanımlanmamış değişken: $var_name");
        }
        print "$var_name = " . $self->{variables}{$var_name} . "\n";
        return $self->{variables}{$var_name};
    }
    
    if ($token =~ /^[a-zA-Z_][a-zA-Z0-9_]*$/) {
        if (@$tokens >= 3 && $tokens->[1] eq '=') {
            my $var_name = $token;
            my $expr_tokens = [@$tokens[2..$#$tokens]];
            my $value = $self->evaluate_expression($expr_tokens);
            $self->{variables}{$var_name} = $value;
            return $value;
        }
        if (exists $self->{variables}{$token}) {
            return $self->{variables}{$token};
        }
        Calculator::Error->throw("Tanımlanmamış değişken: $token");
    }
    
    return $self->evaluate_expression($tokens);
}

sub parse_expression {
    my ($self, $tokens, $pos) = @_;
    return $self->parse_add_sub($tokens, $pos);
}

sub parse_add_sub {
    my ($self, $tokens, $pos) = @_;
    
    my ($left, $new_pos) = $self->parse_mul_div($tokens, $pos);
    
    while ($new_pos < @$tokens && ($tokens->[$new_pos] eq '+' || $tokens->[$new_pos] eq '-')) {
        my $op = $tokens->[$new_pos];
        $new_pos++;
        
        my ($right, $next_pos) = $self->parse_mul_div($tokens, $new_pos);
        
        if ($op eq '+') {
            $left += $right;
        } else {
            $left -= $right;
        }
        
        $new_pos = $next_pos;
    }
    
    return ($left, $new_pos);
}

sub parse_mul_div {
    my ($self, $tokens, $pos) = @_;
    
    my ($left, $new_pos) = $self->parse_primary($tokens, $pos);
    
    while ($new_pos < @$tokens && ($tokens->[$new_pos] eq '*' || $tokens->[$new_pos] eq '/')) {
        my $op = $tokens->[$new_pos];
        $new_pos++;
        
        my ($right, $next_pos) = $self->parse_primary($tokens, $new_pos);
        
        if ($op eq '*') {
            $left *= $right;
        } else {
            if ($right == 0) {
                Calculator::Error->throw("Sıfıra bölme hatası");
            }
            $left /= $right;
        }
        
        $new_pos = $next_pos;
    }
    
    return ($left, $new_pos);
}

sub parse_primary {
    my ($self, $tokens, $pos) = @_;
    
    my $token = $tokens->[$pos];
    
    if ($token =~ /^\d+\.?\d*$/) {
        return ($token, $pos + 1);
    }
    
    if ($token =~ /^[a-zA-Z_][a-zA-Z0-9_]*$/) {
        if (!exists $self->{variables}{$token}) {
            Calculator::Error->throw("Tanımlanmamış değişken: $token");
        }
        return ($self->{variables}{$token}, $pos + 1);
    }
    
    if ($token eq '(') {
        $pos++;
        my ($expr, $new_pos) = $self->parse_expression($tokens, $pos);
        
        if ($new_pos >= @$tokens || $tokens->[$new_pos] ne ')') {
            Calculator::Error->throw("Parantez kapatılmamış");
        }
        
        return ($expr, $new_pos + 1);
    }
    
    Calculator::Error->throw("Geçersiz token: $token");
}

# İşlem önceliği
sub evaluate_expression {
    my ($self, $tokens) = @_;
    return 0 unless @$tokens;
    
    my ($result, $pos) = $self->parse_expression($tokens, 0);
    
    if ($pos < @$tokens) {
        Calculator::Error->throw("İfadenin sonunda beklenmeyen token: " . $tokens->[$pos]);
    }
    
    return $result;
}

package main;

my $calculator = Calculator->new();

while (1) {
    print "> ";
    my $input = <STDIN>;
    last unless defined $input;
    chomp $input;
    
    next if $input eq '';
    
    eval {
        my $tokens = $calculator->tokenize($input);
        my $result = $calculator->evaluate($tokens);
        print "$result\n" unless $input =~ /^print/;
    };
    
    if ($@) {
        if ($@->isa('Calculator::Error')) {
            print "Hata: " . $@->text() . "\n";
        } else {
            print "Beklenmeyen hata: $@\n";
        }
    }
} 

File: rust/calculator.rs
use std::collections::HashMap;
use std::io::{self, Write};
use std::error::Error;
use std::fmt;

#[derive(Debug)]
enum CalculatorError {
    ParseError(String),
    UndefinedVariable(String),
    DivisionByZero,
    InvalidExpression(String),
    IOError(io::Error),
}

impl fmt::Display for CalculatorError {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        match self {
            CalculatorError::ParseError(msg) => write!(f, "Ayrıştırma hatası: {}", msg),
            CalculatorError::UndefinedVariable(name) => write!(f, "Tanımlanmamış değişken: {}", name),
            CalculatorError::DivisionByZero => write!(f, "Sıfıra bölme hatası"),
            CalculatorError::InvalidExpression(msg) => write!(f, "Geçersiz ifade: {}", msg),
            CalculatorError::IOError(e) => write!(f, "Giriş/Çıkış hatası: {}", e),
        }
    }
}

impl Error for CalculatorError {}

impl From<io::Error> for CalculatorError {
    fn from(err: io::Error) -> CalculatorError {
        CalculatorError::IOError(err)
    }
}

#[derive(Debug, Clone)]
enum Token {
    Number(f64),
    Plus,
    Minus,
    Multiply,
    Divide,
    Assign,
    Identifier(String),
    Print,
    LeftParen,
    RightParen,
}

struct Calculator {
    variables: HashMap<String, f64>,
}

impl Calculator {
    fn new() -> Self {
        Calculator {
            variables: HashMap::new(),
        }
    }

    fn evaluate(&mut self, expression: &str) -> Result<f64, CalculatorError> {
        let tokens = self.tokenize(expression)?;
        self.parse_expression(&tokens, 0).map(|(val, _)| val)
    }

    fn tokenize(&self, input: &str) -> Result<Vec<Token>, CalculatorError> {
        let mut tokens = Vec::new();
        let mut chars = input.chars().peekable();

        while let Some(&c) = chars.peek() {
            match c {
                ' ' | '\t' | '\n' => {
                    chars.next();
                    continue;
                }
                '+' => {
                    tokens.push(Token::Plus);
                    chars.next();
                }
                '-' => {
                    tokens.push(Token::Minus);
                    chars.next();
                }
                '*' => {
                    tokens.push(Token::Multiply);
                    chars.next();
                }
                '/' => {
                    tokens.push(Token::Divide);
                    chars.next();
                }
                '=' => {
                    tokens.push(Token::Assign);
                    chars.next();
                }
                '(' => {
                    tokens.push(Token::LeftParen);
                    chars.next();
                }
                ')' => {
                    tokens.push(Token::RightParen);
                    chars.next();
                }
                c if c.is_alphabetic() => {
                    let mut identifier = String::new();
                    while let Some(&c) = chars.peek() {
                        if c.is_alphanumeric() {
                            identifier.push(c);
                            chars.next();
                        } else {
                            break;
                        }
                    }
                    if identifier == "print" {
                        tokens.push(Token::Print);
                    } else {
                        tokens.push(Token::Identifier(identifier));
                    }
                }
                c if c.is_digit(10) || c == '.' => {
                    let mut number = String::new();
                    while let Some(&c) = chars.peek() {
                        if c.is_digit(10) || c == '.' {
                            number.push(c);
                            chars.next();
                        } else {
                            break;
                        }
                    }
                    match number.parse::<f64>() {
                        Ok(n) => tokens.push(Token::Number(n)),
                        Err(_) => return Err(CalculatorError::ParseError("Geçersiz sayı formatı".to_string())),
                    }
                }
                _ => return Err(CalculatorError::ParseError(format!("Geçersiz karakter: {}", c))),
            }
        }
        Ok(tokens)
    }

    // Yeni recursive descent parsing fonksiyonları
    fn parse_expression(&mut self, tokens: &[Token], pos: usize) -> Result<(f64, usize), CalculatorError> {
        if tokens.is_empty() {
            return Err(CalculatorError::InvalidExpression("Boş ifade".to_string()));
        }

        let (mut left, mut new_pos) = self.parse_term(tokens, pos)?;

        while new_pos < tokens.len() {
            match &tokens[new_pos] {
                Token::Plus => {
                    let (right, right_pos) = self.parse_term(tokens, new_pos + 1)?;
                    left += right;
                    new_pos = right_pos;
                },
                Token::Minus => {
                    let (right, right_pos) = self.parse_term(tokens, new_pos + 1)?;
                    left -= right;
                    new_pos = right_pos;
                },
                _ => break,
            }
        }

        Ok((left, new_pos))
    }

    fn parse_term(&mut self, tokens: &[Token], pos: usize) -> Result<(f64, usize), CalculatorError> {
        if pos >= tokens.len() {
            return Err(CalculatorError::InvalidExpression("Beklenmeyen ifade sonu".to_string()));
        }

        let (mut left, mut new_pos) = self.parse_factor(tokens, pos)?;

        while new_pos < tokens.len() {
            match &tokens[new_pos] {
                Token::Multiply => {
                    let (right, right_pos) = self.parse_factor(tokens, new_pos + 1)?;
                    left *= right;
                    new_pos = right_pos;
                },
                Token::Divide => {
                    let (right, right_pos) = self.parse_factor(tokens, new_pos + 1)?;
                    if right == 0.0 {
                        return Err(CalculatorError::DivisionByZero);
                    }
                    left /= right;
                    new_pos = right_pos;
                },
                _ => break,
            }
        }

        Ok((left, new_pos))
    }

    fn parse_factor(&mut self, tokens: &[Token], pos: usize) -> Result<(f64, usize), CalculatorError> {
        if pos >= tokens.len() {
            return Err(CalculatorError::InvalidExpression("Beklenmeyen ifade sonu".to_string()));
        }

        match &tokens[pos] {
            Token::Number(n) => Ok((*n, pos + 1)),
            Token::Identifier(name) => {
                // İşlem atama ise
                if pos + 1 < tokens.len() && matches!(tokens[pos + 1], Token::Assign) {
                    let (value, new_pos) = self.parse_expression(tokens, pos + 2)?;
                    self.variables.insert(name.clone(), value);
                    Ok((value, new_pos))
                } else if let Some(&value) = self.variables.get(name) {
                    Ok((value, pos + 1))
                } else {
                    Err(CalculatorError::UndefinedVariable(name.clone()))
                }
            },
            Token::Print => {
                if pos + 1 >= tokens.len() {
                    return Err(CalculatorError::InvalidExpression("Print ifadesi için değer gerekli".to_string()));
                }
                let (value, new_pos) = self.parse_expression(tokens, pos + 1)?;
                println!("{}", value);
                Ok((value, new_pos))
            },
            Token::LeftParen => {
                let (value, new_pos) = self.parse_expression(tokens, pos + 1)?;
                if new_pos >= tokens.len() || !matches!(tokens[new_pos], Token::RightParen) {
                    return Err(CalculatorError::InvalidExpression("Kapanmamış parantez".to_string()));
                }
                Ok((value, new_pos + 1))
            },
            _ => Err(CalculatorError::InvalidExpression("Geçersiz ifade başlangıcı".to_string())),
        }
    }
}

fn main() -> Result<(), Box<dyn Error>> {
    let mut calculator = Calculator::new();
    println!("Rust Hesap Makinesi");
    println!("Çıkmak için 'exit' yazın");

    loop {
        print!("> ");
        io::stdout().flush()?;
        let mut input = String::new();
        io::stdin().read_line(&mut input)?;

        let input = input.trim();
        if input.is_empty() {
            continue;
        }
        if input == "exit" {
            break;
        }

        match calculator.evaluate(input) {
            Ok(result) => println!("{}", result),
            Err(e) => println!("Hata: {}", e),
        }
    }

    Ok(())
}
